### DISCUSSION TITLE
Use private git repository for build context flag

### DISCUSSION THREAD

**Discussion by OttoHung** [docker/buildx#1228] [Q&A]
Hi,
Based on [build context](https://docs.docker.com/engine/reference/commandline/buildx_build/#build-context) page, the `Buildx` can download a remote git repository to the local device as the build context. This is working while assigning a public git repository to the build context flag, but it doesn't work when the git repository is private.

I wrote a [script](https://github.com/OttoHung/docker.tutorials/blob/main/scripts/build-contexts/git/https/build_from_private.sh) to test this scenario, then I got an error message, `fatal: could not read Username for 'https://github.com': terminal prompts disabled`, as the response.

I would like to know if there is a way to download the private git repository via the build context flag?

Thank you very much.

Have a good one.

**Comment by jedevc** [comment]
Heya :wave:

Have you tried a clone over the `ssh://` protocol instead of `https://`? https basic auth is not supported, but you should able to easily use the ssh option. 

If you have problems, the instructions and example in the [buildx build](https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#-ssh-agent-socket-or-keys-to-expose-to-the-build---ssh) docs should be helpful.

**Reply by OttoHung** [reply]
Hi @jedevc ,
Thank you for the quick reply. I tried to clone over the `ssh://` then I got `error: failed to get build context repo: stat ssh://git@github.com:ottohung/docker.tutorial.git#main: no such file or directory`.

If I use `git@github.com:ottohung/docker.tutorial.git#main` to clone the repo, then I got `error: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to solve with frontend gateway.v0: rpc error: code = Unknown desc = unsupported context source git@github.com for repo`

The command prompt is as follows:
```shell
REPO=ssh://git@github.com:ottohung/docker.tutorial.git#main
docker buildx build \
   --build-context repo=${REPO} \
   --secret id=npm,src=$HOME/.npmrc \
   --ssh default=$SSH_AUTH_SOCK \
   --no-cache \
   -t ${IMG_NAME} \
   -f ./dockerfiles/build-contexts/git/ssh/Dockerfile .
```

The `dockerfile` is as follows:
```dockerfile
# syntax=docker/dockerfile:1.4

FROM node:14.17.1-alpine AS sourceStage
WORKDIR /source

RUN apk update && apk upgrade \
    apk add --no-cache openssh-client git

RUN mkdir -p -m 0700 ~/.ssh \
    && ssh-keyscan -H github.com >> ~/.ssh/known_hosts

COPY --from=repo ["./", "/source"]

FROM node:14.17.1-alpine AS buildStage
WORKDIR /build

COPY --from=sourceStage ["/source", "/build"]
RUN --mount=type=secret,id=npm,target=/build/.npmrc \
    yarn install
RUN yarn build


FROM node:14.17.1-alpine AS moduleStage
WORKDIR /module
COPY --from=buildStage ["/build/node_modules", "/module/node_modules"]
COPY --from=buildStage ["/build/package.json", "/module"]
COPY --from=buildStage ["/build/LICENSE", "/module"]

COPY --from=buildStage ["/build/workspaces/greeting/dist", "/module/workspaces/greeting/dist"]

COPY --from=buildStage ["/build/workspaces/greeting/package.json", "/module/workspaces/greeting"]

FROM node:14.17.1-alpine AS appStage
WORKDIR /app
COPY --from=moduleStage ["/module", "/app"]

EXPOSE 5478

ENTRYPOINT [ "node", "workspaces/greeting/dist/server.js"]
```

I tried to use `https://${PAT}@github.com/OttoHung/private-test.git#main` to clone a private git repo and this works.
